Master Big Data et Intelligence Artificielle
Nom de l'étudiant: AZIZ BOUFRIOUA                 
                                                  Examen de Machine Learning 1

 Mr. Youssef ES-SAADY et Mohamed El HAJJI

1 - Analysez les données et proposer une maquette IHM

# Anayse et prétraitement des données

In [33]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler

# Models
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC, SVC
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier


import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

Nous devons importer les données pour faire une analyse et un prétraitement si c'est néssaicaire 

In [34]:
# Chargement des données
data = pd.read_csv("diabetes_data_upload.csv") 

In [35]:
#Affichage de data
data

Age  Gender Polyuria Polydipsia sudden weight loss weakness Polyphagia  \
0     40    Male       No        Yes                 No      Yes         No   
1     58    Male       No         No                 No      Yes         No   
2     41    Male      Yes         No                 No      Yes        Yes   
3     45    Male       No         No                Yes      Yes        Yes   
4     60    Male      Yes        Yes                Yes      Yes        Yes   
..   ...     ...      ...        ...                ...      ...        ...   
515   39  Female      Yes        Yes                Yes       No        Yes   
516   48  Female      Yes        Yes                Yes      Yes        Yes   
517   58  Female      Yes        Yes                Yes      Yes        Yes   
518   32  Female       No         No                 No      Yes         No   
519   42    Male       No         No                 No       No         No   

    Genital thrush visual blurring Itching Irritability delayed healing  \
0               No              No     Yes           No             Yes   
1               No             Yes      No           No              No   
2               No              No     Yes           No             Yes   
3              Yes              No     Yes           No             Yes   
4               No             Yes     Yes          Yes             Yes   
..             ...             ...     ...          ...             ...   
515             No              No     Yes           No             Yes   
516             No              No     Yes          Yes             Yes   
517             No             Yes      No           No              No   
518             No             Yes     Yes           No             Yes   
519             No              No      No           No              No   

    partial paresis muscle stiffness Alopecia Obesity     class  
0                No              Yes      Yes     Yes  Positive  
1               Yes               No      Yes      No  Positive  
2                No              Yes      Yes      No  Positive  
3                No               No       No      No  Positive  
4               Yes              Yes      Yes     Yes  Positive  
..              ...              ...      ...     ...       ...  
515             Yes               No       No      No  Positive  
516             Yes               No       No      No  Positive  
517             Yes              Yes       No     Yes  Positive  
518              No               No      Yes      No  Negative  
519              No               No       No      No  Negative  

[520 rows x 17 columns]

In [36]:
#Verifier si on a pas de valeur manquantes 
data.isna().sum()


Age                   0
Gender                0
Polyuria              0
Polydipsia            0
sudden weight loss    0
weakness              0
Polyphagia            0
Genital thrush        0
visual blurring       0
Itching               0
Irritability          0
delayed healing       0
partial paresis       0
muscle stiffness      0
Alopecia              0
Obesity               0
class                 0
dtype: int64

on a pas de valeur manquantes 

In [37]:
#avoir des information sur les type de chaque colonne
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 520 entries, 0 to 519
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Age                 520 non-null    int64 
 1   Gender              520 non-null    object
 2   Polyuria            520 non-null    object
 3   Polydipsia          520 non-null    object
 4   sudden weight loss  520 non-null    object
 5   weakness            520 non-null    object
 6   Polyphagia          520 non-null    object
 7   Genital thrush      520 non-null    object
 8   visual blurring     520 non-null    object
 9   Itching             520 non-null    object
 10  Irritability        520 non-null    object
 11  delayed healing     520 non-null    object
 12  partial paresis     520 non-null    object
 13  muscle stiffness    520 non-null    object
 14  Alopecia            520 non-null    object
 15  Obesity             520 non-null    object
 16  class               520 no

In [38]:
# Les valeurs uniques de chaque colonne
for column_name in data.columns:
    unique_values = data[column_name].unique()
    print(f"Valeurs uniques dans la colonne '{column_name}': {unique_values}")


Valeurs uniques dans la colonne 'Age': [40 58 41 45 60 55 57 66 67 70 44 38 35 61 54 43 62 39 48 32 42 52 53 37
 49 63 30 50 46 36 51 59 65 25 47 28 68 56 31 85 90 72 69 79 34 16 33 64
 27 29 26]
Valeurs uniques dans la colonne 'Gender': ['Male' 'Female']
Valeurs uniques dans la colonne 'Polyuria': ['No' 'Yes']
Valeurs uniques dans la colonne 'Polydipsia': ['Yes' 'No']
Valeurs uniques dans la colonne 'sudden weight loss': ['No' 'Yes']
Valeurs uniques dans la colonne 'weakness': ['Yes' 'No']
Valeurs uniques dans la colonne 'Polyphagia': ['No' 'Yes']
Valeurs uniques dans la colonne 'Genital thrush': ['No' 'Yes']
Valeurs uniques dans la colonne 'visual blurring': ['No' 'Yes']
Valeurs uniques dans la colonne 'Itching': ['Yes' 'No']
Valeurs uniques dans la colonne 'Irritability': ['No' 'Yes']
Valeurs uniques dans la colonne 'delayed healing': ['Yes' 'No']
Valeurs uniques dans la colonne 'partial paresis': ['No' 'Yes']
Valeurs uniques dans la colonne 'muscle stiffness': ['Yes' 'No']
Valeurs 

Dans notre cas d'étude on a une base de données qui donne un ensemble d'information sur des personnes atteint ou diagnostiquer de diabète comme le montre la colonne classe : negatif ou positif ainsi q'un nombre d'information tel que Âge,Genre,Polyurie (augmentation de la production d’urine),Polydipsie (soif excessive),Perte de poids soudaine,Faiblesse,Polyphagie (augmentation de l’appétit)...Alopécie (perte de cheveux)' c'est information dites variable explicative permettrons de predir la probabilité d'etre diagnostiquer comme positif ou non, donc on peut en déduire que nous allons utiliser des modèles de classification binaires.

et vue que nous aurons plusieur variable a entré pour la prédiction je propose l'IHM suivante:


![Description de l'image](Screenshot1.png)
![Description de l'image](Screenshot2.png)



l'appliacation va peremtre au utilisateurs de renter tous les données (variable explicative) de l'age à l'obésité, puis il va cliquer sur un bouton Prédire pour que le model prédie la classe positif o négatif et un message s'affichera comme suite:
<br>


Sauf l'age tous les variables ont deux valeurs uniques (comme 'Yes' et 'No' ou 'Male' et 'Female'), nous pouvons donc utiliser Label Encoding ou Binary Encoding. 
Pour la variable 'Age'c'est une variable quantitative donc on va la standarisé

In [39]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Supposez que votre dataset est dans une DataFrame nommée `data`

# Encodage pour les variables binaires
binary_columns = ['Gender', 'Polyuria', 'Polydipsia', 'sudden weight loss', 'weakness', 'Polyphagia', 'Genital thrush', 
                  'visual blurring', 'Itching', 'Irritability', 'delayed healing', 'partial paresis', 'muscle stiffness', 
                  'Alopecia', 'Obesity', 'class']

le = LabelEncoder()
for col in binary_columns:
    data[col] = le.fit_transform(data[col])

# Standard Scaling pour la variable Age
scaler = StandardScaler()
data['Age'] = scaler.fit_transform(data[['Age']])

# Vérifiez le résultat final
print(data.head())


import joblib
#saufgarder le coder et le sclar
joblib.dump(le, 'encoder_diabete.pkl')
joblib.dump(scaler, 'scaler_diabete.pkl')


        Age  Gender  Polyuria  Polydipsia  sudden weight loss  weakness  \
0 -0.661367       1         0           1                   0         1   
1  0.821362       1         0           0                   0         1   
2 -0.578993       1         1           0                   0         1   
3 -0.249498       1         0           0                   1         1   
4  0.986110       1         1           1                   1         1   

   Polyphagia  Genital thrush  visual blurring  Itching  Irritability  \
0           0               0                0        1             0   
1           0               0                1        0             0   
2           1               0                0        1             0   
3           1               1                0        1             0   
4           1               0                1        1             1   

   delayed healing  partial paresis  muscle stiffness  Alopecia  Obesity  \
0                1                

['scaler_diabete.pkl']

In [40]:
# Les valeurs uniques de chaque colonne
for column_name in data.columns:
    unique_values = data[column_name].unique()
    print(f"Valeurs uniques dans la colonne '{column_name}': {unique_values}")

Valeurs uniques dans la colonne 'Age': [-6.61366896e-01  8.21362241e-01 -5.78993055e-01 -2.49497691e-01
  9.86109923e-01  5.74240718e-01  7.38988400e-01  1.48035297e+00
  1.56272681e+00  1.80984833e+00 -3.31871532e-01 -8.26114578e-01
 -1.07323610e+00  1.06848376e+00  4.91866877e-01 -4.14245373e-01
  1.15085761e+00 -7.43740737e-01 -2.37616849e-03 -1.32035762e+00
 -4.96619214e-01  3.27119195e-01  4.09493036e-01 -9.08488419e-01
  7.99976725e-02  1.23323145e+00 -1.48510531e+00  1.62371513e-01
 -1.67123850e-01 -9.90862260e-01  2.44745354e-01  9.03736082e-01
  1.39797913e+00 -1.89697451e+00 -8.47500095e-02 -1.64985299e+00
  1.64510065e+00  6.56614559e-01 -1.40273147e+00  3.04545595e+00
  3.45732515e+00  1.97459601e+00  1.72747449e+00  2.55121290e+00
 -1.15560994e+00 -2.63833908e+00 -1.23798378e+00  1.31560529e+00
 -1.73222683e+00 -1.56747915e+00 -1.81460067e+00]
Valeurs uniques dans la colonne 'Gender': [1 0]
Valeurs uniques dans la colonne 'Polyuria': [0 1]
Valeurs uniques dans la colonne '

In [41]:
data.describe()

Age      Gender    Polyuria  Polydipsia  sudden weight loss  \
count  5.200000e+02  520.000000  520.000000  520.000000          520.000000   
mean   5.807320e-17    0.630769    0.496154    0.448077            0.417308   
std    1.000963e+00    0.483061    0.500467    0.497776            0.493589   
min   -2.638339e+00    0.000000    0.000000    0.000000            0.000000   
25%   -7.437407e-01    0.000000    0.000000    0.000000            0.000000   
50%   -4.356309e-02    1.000000    0.000000    0.000000            0.000000   
75%    7.389884e-01    1.000000    1.000000    1.000000            1.000000   
max    3.457325e+00    1.000000    1.000000    1.000000            1.000000   

         weakness  Polyphagia  Genital thrush  visual blurring     Itching  \
count  520.000000  520.000000      520.000000       520.000000  520.000000   
mean     0.586538    0.455769        0.223077         0.448077    0.486538   
std      0.492928    0.498519        0.416710         0.497776    0.500300   
min      0.000000    0.000000        0.000000         0.000000    0.000000   
25%      0.000000    0.000000        0.000000         0.000000    0.000000   
50%      1.000000    0.000000        0.000000         0.000000    0.000000   
75%      1.000000    1.000000        0.000000         1.000000    1.000000   
max      1.000000    1.000000        1.000000         1.000000    1.000000   

       Irritability  delayed healing  partial paresis  muscle stiffness  \
count    520.000000       520.000000       520.000000        520.000000   
mean       0.242308         0.459615         0.430769          0.375000   
std        0.428892         0.498846         0.495661          0.484589   
min        0.000000         0.000000         0.000000          0.000000   
25%        0.000000         0.000000         0.000000          0.000000   
50%        0.000000         0.000000         0.000000          0.000000   
75%        0.000000         1.000000         1.000000          1.000000   
max        1.000000         1.000000         1.000000          1.000000   

         Alopecia     Obesity       class  
count  520.000000  520.000000  520.000000  
mean     0.344231    0.169231    0.615385  
std      0.475574    0.375317    0.486973  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    1.000000  
75%      1.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000

### Décrivez la méthodologie, les algorithmes, et les métriques d'évaluation à utiliser.


Pour la méthodologie utilisé, j'ai essayer de comprendre le jeu de données a partir de ces colonnes et leur valeurs, la premère remarque été le nombre d'enregistrement qui ne depasse pas 520, donc automatiquement j'ai pensé a faire une validation dymamique.
et puisque tout les variables n'ont que deux attribue sauf l'âge donc j'ai pensé a utilisé un scaler pour le standarisé et un labelencoder pour les autres.
pour les alogorithmes utilisés :

1. Logistic Regression
La régression logistique est un modèle de classification linéaire qui prédit la probabilité qu'un échantillon appartient à une certaine classe. Elle utilise la fonction logistique (sigmoïde) pour transformer les valeurs de sortie en probabilités. Elle est souvent utilisée pour les problèmes de classification binaire.

2. K-Nearest Neighbors (KNN)
KNN est un algorithme non paramétrique de classification qui attribue une classe à un échantillon en fonction des classes des K voisins les plus proches dans l'espace des caractéristiques. Il est basé sur la distance, généralement la distance euclidienne.

3. Decision Tree
Les arbres de décision sont des modèles de classification ou de régression qui partitionnent les données en sous-ensembles basés sur les valeurs des caractéristiques. Chaque noeud interne représente un test sur une caractéristique, chaque branche représente le résultat du test, et chaque feuille représente une classe ou une valeur de régression.

4. Support Vector Machine (SVM) avec noyau RBF
SVM est un algorithme de classification qui cherche à trouver l'hyperplan optimal séparant les différentes classes dans un espace de caractéristiques. Le noyau RBF (Radial Basis Function) permet de transformer les données non linéairement séparables en un espace de caractéristiques plus élevé où elles deviennent linéairement séparables.

5. Random Forest
Les forêts aléatoires sont un ensemble d'arbres de décision construits sur des échantillons aléatoires du jeu de données et des sous-ensembles aléatoires des caractéristiques. La classification est déterminée par la majorité des votes des arbres individuels, ce qui réduit le risque de surapprentissage.

Pour les métriques :
 matrice de confusion 
 ![matrice de confusion](Screenshot8.png)
 <br>
Le rappel ("recall"  en anglais), ou sensibilité ("sensitivity" en anglais), est le taux de vrais positifs, c’est à dire la proportion de positifs que l’on a correctement identifiés. C’est la capacité de notre modèle à détecter tous les positifs dans notre cas.
<br>
![matrice de confusion](Screenshot4.png)
 



On s’intéressera donc aussi à la précision, c’est-à-dire la proportion de prédictions correctes parmi les points que l’on a prédits positifs. C’est la capacité de notre modèle à ne detecter que pour un vrai positif</br>
![Example Image](Screenshot5.png)



Pour évaluer un compromis entre rappel et précision, on peut calculer la "F-mesure", qui est leur moyenne harmonique.</br>
![Example Image](Screenshot6.png)

Pour finir, on s’intéresse aussi souvent à la spécificité ("specificity" en anglais), qui est le taux de vrais négatifs, autrement dit la capacité à détecter toutes les  situations où il n’y a pas de positif. C’est une mesure complémentaire de la sensibilité.</br>
![Example Image](Screenshot7.png)

Nous verifions si les données sont bien encoder et standardisé, nous avons 520 entrée en somme donc je suppose q'il vaut mieux faire une validation croisé :stratégie et évaluation sur les petits échantillons

In [42]:
# Créer une nouvelle variable 'X' qui contient uniquement les colonnes 'gender', 'ethnicity', 'parent_education', 'lunch', et 'prep' du DataFrame 'df'
X = data[['Age','Gender', 'Polyuria', 'Polydipsia', 'sudden weight loss', 'weakness', 'Polyphagia', 'Genital thrush', 
                  'visual blurring', 'Itching', 'Irritability', 'delayed healing', 'partial paresis', 'muscle stiffness', 
                  'Alopecia', 'Obesity']]

# Créer une nouvelle variable 'y' qui contient uniquement la colonne 'Result' du DataFrame 'df'
y = data['class']

In [43]:
X

Age  Gender  Polyuria  Polydipsia  sudden weight loss  weakness  \
0   -0.661367       1         0           1                   0         1   
1    0.821362       1         0           0                   0         1   
2   -0.578993       1         1           0                   0         1   
3   -0.249498       1         0           0                   1         1   
4    0.986110       1         1           1                   1         1   
..        ...     ...       ...         ...                 ...       ...   
515 -0.743741       0         1           1                   1         0   
516 -0.002376       0         1           1                   1         1   
517  0.821362       0         1           1                   1         1   
518 -1.320358       0         0           0                   0         1   
519 -0.496619       1         0           0                   0         0   

     Polyphagia  Genital thrush  visual blurring  Itching  Irritability  \
0             0               0                0        1             0   
1             0               0                1        0             0   
2             1               0                0        1             0   
3             1               1                0        1             0   
4             1               0                1        1             1   
..          ...             ...              ...      ...           ...   
515           1               0                0        1             0   
516           1               0                0        1             1   
517           1               0                1        0             0   
518           0               0                1        1             0   
519           0               0                0        0             0   

     delayed healing  partial paresis  muscle stiffness  Alopecia  Obesity  
0                  1                0                 1         1        1  
1                  0                1                 0         1        0  
2                  1                0                 1         1        0  
3                  1                0                 0         0        0  
4                  1                1                 1         1        1  
..               ...              ...               ...       ...      ...  
515                1                1                 0         0        0  
516                1                1                 0         0        0  
517                0                1                 1         0        1  
518                1                0                 0         1        0  
519                0                0                 0         0        0  

[520 rows x 16 columns]

In [44]:
y

0      1
1      1
2      1
3      1
4      1
      ..
515    1
516    1
517    1
518    0
519    0
Name: class, Length: 520, dtype: int32

In [45]:
# Utiliser la fonction 'train_test_split' pour diviser les données en un ensemble d'entraînement et un ensemble de test
# 'X' et 'y' sont les caractéristiques et les cibles, respectivement
# 'test_size = 0.2' signifie que 20% des données seront utilisées pour le test
# 'random_state = 42' est utilisé pour assurer la reproductibilité des résultats
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [46]:
from sklearn.metrics import recall_score, precision_score, f1_score, confusion_matrix


models = {
    "                   Logistic Regression": LogisticRegression(),
    "                   K-Nearest Neighbors": KNeighborsClassifier(),
    "                         Decision Tree": DecisionTreeClassifier(),
    "   Support Vector Machine (RBF Kernel)": SVC(),
    "                         Random Forest": RandomForestClassifier(),
}

for name, model in models.items():
    model.fit(X_train, y_train)
    print(name + " trained.")



    # Prédire les étiquettes de l'ensemble de test
    y_pred = model.predict(X_test)

    # Calculer les métriques
    recall = recall_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    f_score = f1_score(y_test, y_pred)
    tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
    specificity = tn / (tn+fp)

    # Afficher les métriques
    print(f"Model: {model.__class__.__name__}")
    print(f"Recall: {recall}")
    print(f"Precision: {precision}")
    print(f"F-score: {f_score}")
    print(f"Specificity: {specificity}")
    print("------------------------")

                   Logistic Regression trained.
Model: LogisticRegression
Recall: 0.9577464788732394
Precision: 0.9315068493150684
F-score: 0.9444444444444444
Specificity: 0.8484848484848485
------------------------
                   K-Nearest Neighbors trained.
Model: KNeighborsClassifier
Recall: 0.9014084507042254
Precision: 0.9552238805970149
F-score: 0.927536231884058
Specificity: 0.9090909090909091
------------------------
                         Decision Tree trained.
Model: DecisionTreeClassifier
Recall: 0.971830985915493
Precision: 1.0
F-score: 0.9857142857142858
Specificity: 1.0
------------------------
   Support Vector Machine (RBF Kernel) trained.
Model: SVC
Recall: 0.9859154929577465
Precision: 0.9722222222222222
F-score: 0.979020979020979
Specificity: 0.9393939393939394
------------------------
                         Random Forest trained.
Model: RandomForestClassifier
Recall: 0.9859154929577465
Precision: 1.0
F-score: 0.9929078014184397
Specificity: 1.0
-------------

Analyse
Recall : Important pour minimiser les faux négatifs. Les modèles SVM et Random Forest ont les meilleurs résultats (0.9859).
Precision : Important pour minimiser les faux positifs. Les modèles Decision Tree et Random Forest ont la meilleure précision (1.0).
F-score : Combine à la fois la précision et le rappel. Le modèle Random Forest a le meilleur F-score (0.9929).
Specificity : Important pour minimiser les faux positifs parmi les négatifs. Les modèles Decision Tree et Random Forest ont la meilleure spécificité (1.0).
Conclusion
Le modèle Random Forest semble être le meilleur modèle dans notre cas, car il a les meilleurs résultats pour la plupart des métriques importantes :

Plus haut rappel (0.9859)
Plus haute précision (1.0)
Meilleur F-score (0.9929)
Plus haute spécificité (1.0)

In [47]:
# Evaluation des models avec la validation croisé
for name, model in models.items():
    scores = cross_val_score(model, X_train, y_train, scoring='accuracy', cv=5)
    print(name, "average score: {:.2f} %".format(scores.mean() * 100))

                   Logistic Regression average score: 91.58 %
                   K-Nearest Neighbors average score: 92.31 %
                         Decision Tree average score: 95.92 %
   Support Vector Machine (RBF Kernel) average score: 93.51 %
                         Random Forest average score: 97.12 %


In [48]:
# Prédire sur l'ensemble de test
for name, model in models.items():
    print(name + "score on test set: {:.2f} %".format(model.score(X_test, y_test) * 100))

                   Logistic Regressionscore on test set: 92.31 %
                   K-Nearest Neighborsscore on test set: 90.38 %
                         Decision Treescore on test set: 98.08 %
   Support Vector Machine (RBF Kernel)score on test set: 97.12 %
                         Random Forestscore on test set: 99.04 %


Sur la base des résultats de la validation croisée, le modèle avec la meilleure performance en termes de précision sur le jeu de données est le Random Forest avec un score de 99.04%.

Avantages du modèle Random Forest
Robustesse : Les forêts aléatoires combinent les résultats de plusieurs arbres de décision, réduisant le risque de surapprentissage (overfitting).
Performance élevée : Les résultats montrent une haute précision et de bonnes métriques de performance.
Flexibilité : Les forêts aléatoires peuvent bien gérer les jeux de données complexes et les interactions entre les caractéristiques.